### PR TITLE
feat(admin): hold and renew a document claim while editing

### PR DESCRIPTION
**Pull request by mobeenabdullah**
G20, the admin data layer. No visual change yet: this is the hook the banner will read.

## What it does

Claims on mount, renews on the heartbeat, releases on unmount, and reports who holds the document when someone else does. Advisory throughout, so nothing here refuses a write and the server refuses none either. What it buys is that two editors are **told** about each other rather than discovering it when one overwrites the other.

## Four behaviours that would fail silently

Each is pinned by a test that fails when the behaviour is removed.

- **A release only fires for a claim this editor actually held.** Releasing a colleague's would hand the document to whoever looked at it next.
- **Renewal stops once a takeover is reported.** Asking again would either fail forever or re-take a claim the editor was just told they had lost. The outcome carries the new holder, so the UI can name *who*.
- **A failed renew is a blip, not a lost claim.** The lease outlives several beats, so a dropped packet resolves itself on the next one; treating it as a takeover would move an editor to read-only over one lost request.
- **The claim happens once per document, not once per render.** The document reference is memoised on the three values that identify it; rebuilt each render, the effect would release and re-claim on every one, and a colleague would watch the lock flicker between held and free.

React mounts effects twice in development, so a naive version claims, releases and claims again, and the first cleanup can land after the second claim. Each run carries a generation and a reply belonging to a superseded one is dropped. Release is token-scoped on the server as well, which is the same guarantee arrived at from the other side.

## Why `nextly/document-lock` exists

The heartbeat is the agreement between a lease and whoever renews it, and a second copy of the number drifts from the first the moment either is tuned, so the hook imports it rather than choosing one.

It is its own client entry rather than part of the root, and that was measured rather than assumed. `packages/admin/tsconfig.json` maps `nextly` to this package's **source**, so reaching two constants through the root entry pulls the DI container and the auth middleware into the admin's typecheck: **112 errors, none of them about the importing code**. The precedent is already in that file — `nextly/widget-gate` exists for the same reason, with the rationale written beside it.

The entry carries the lease timings and the wire types and nothing else. Verified in the **built artefacts** rather than the source: `dist/document-lock.d.ts` contains 8 contract symbols and **0** repository or adapter symbols, so the thing that takes a `DrizzleAdapter` stays behind the route where it belongs.

## Two false signals met on the way

Worth recording, because the obvious reading of either is that your own change caused it.

The admin's typecheck first reported **121 errors, none mine**, from a `nextly` dist older than the per-widget-settings merge. Rebuilding cleared them. An error count that says nothing about the code you changed is the same trap as a green that says nothing.

And the first push was **rejected by the pre-push hook** over two `react-hooks/exhaustive-deps` warnings in this file, because the admin's lint script runs `--max-warnings 0`. Checking with bare `eslint` reported "0 errors, 2 warnings" and looked clean. The warnings were real, and fixing them properly is the memoisation described above.

## Verification

11 hook tests. All four failure modes mutation-tested. Both packages typecheck clean on their own exit codes, `pnpm lint:design`, `pnpm lint:workspace` and `turbo lint` all pass, and the comment-convention and docs-claims guards pass.

## Next

The visual layer: a banner naming the holder with take-over, read-only rendering through the existing `ReadOnlyDocumentForm` contract, and lock status joining `DocumentStatusLive` rather than adding a second live region — that component's own comment warns that two live regions in one header interrupt each other.

### CODE REVIEW

File: packages/admin/src/hooks/queries/useDocumentLock.ts [hunk 1]
```diff
@@ -0,0 +1,163 @@
...
+  /** Off for a read-only surface: a past version is not being edited. */
+  enabled?: boolean;
+}
+
+interface MutationEnvelope<T> {
+  message: string;
+  item: T;
+}
+
+export function useDocumentLock({
```
[the remark this answers is not in the document]
Comment (mobeenabdullah) [reply packages/admin/src/hooks/queries/useDocumentLock.ts:126]: Right that nothing mounts it - it is exported from no barrel and absent from the admin's export map, so nothing an installer of this version can observe changes. That is deliberate: this PR is the data layer, and the editor surface is the next one, so that the banner, the read-only rendering and the take-over affordance get reviewed as a unit rather than bolted onto a data-layer diff.

The real defect you found is the changeset, which claimed the shipped behaviour. It now describes what actually ships - the client entry, the module split and the claim mechanics - and says plainly that no editor mounts the hook yet.

File: packages/admin/src/hooks/queries/useDocumentLock.ts [hunk 2]
```diff
@@ -0,0 +1,163 @@
...
+  /** Absent while a document is being created, which is nothing to claim. */
+  entryId?: string | null;
+  /** Off for a read-only surface: a past version is not being edited. */
+  enabled?: boolean;
+}
+
+interface MutationEnvelope<T> {
+  message: string;
+  item: T;
+}
```
[the remark this answers is not in the document]
Comment (mobeenabdullah) [reply packages/admin/src/hooks/queries/useDocumentLock.ts]: Fixed. `MutationEnvelope` is deleted and both calls use `MutationResponse<T>` from `@admin/lib/api/response-types`.

File: packages/admin/src/hooks/queries/useDocumentLock.ts [hunk 3]
```diff
@@ -0,0 +1,163 @@
...
+          // as a takeover would move an editor to read-only for a dropped
+          // packet.
+        });
+    }, DOCUMENT_LOCK_HEARTBEAT_INTERVAL_MS);
+
+    return () => {
+      clearInterval(heartbeat);
+      const token = claimToken.current;
+      claimToken.current = null;
+      if (!token) return;
```
[the remark this answers is not in the document]
Comment (mobeenabdullah) [reply packages/admin/src/hooks/queries/useDocumentLock.ts]: Fixed. Cleanup sets a per-run `cancelled` flag, and an acquire that resolves afterwards releases the claim it was handed instead of storing it. Without it a tab closed during the acquire left a claim nothing would renew or release, and the next editor waited out a whole lease rather than one request. Test settles the acquire after `unmount()` and asserts the DELETE.

File: packages/admin/src/hooks/queries/useDocumentLock.ts [hunk 4]
```diff
@@ -0,0 +1,163 @@
...
+          claimToken: claimToken.current,
+        })
+        .then(({ item }) => {
+          if (run !== generation.current || item.status === "renewed") return;
+          // Taken over. Stop renewing and say who has it: asking again would
+          // either fail forever or re-take a claim just reported as lost.
+          claimToken.current = null;
+          setState({ status: "taken-over", holder: item.holder });
+        })
+        .catch(() => {
```
[the remark this answers is not in the document]
Comment (mobeenabdullah) [reply packages/admin/src/hooks/queries/useDocumentLock.ts:309]: Fixed. `DOCUMENT_LOCK_LOSS_AFTER_MS` was exported for exactly this and the hook never imported it. A failed renewal stays a blip until the deadline and reports `lost` past it, when the lease has expired server-side and a colleague may already hold the row. Two tests, deliberately in both directions: one failure inside the deadline keeps `held-by-me`, and a run of them past it reaches `lost` and stops the beat. A mutation that reports loss on the first failure fails the first test.

File: packages/admin/src/hooks/queries/useDocumentLock.ts [hunk 5]
```diff
@@ -0,0 +1,163 @@
...
+
+  useEffect(() => {
+    if (!active) {
+      setState({ status: "idle" });
+      return;
+    }
+
+    generation.current += 1;
+    const run = generation.current;
+    void claim(false, run);
```
[the remark this answers is not in the document]
Comment (mobeenabdullah) [reply packages/admin/src/hooks/queries/useDocumentLock.ts]: Fixed. The acquire's rejection is caught, reported as `unavailable`, and retried on the next beat, which now re-acquires whenever there is no token. You are right that it was silent for the whole session otherwise: every beat exited at the token guard. Test rejects the first POST, asserts `unavailable`, then lets the next beat succeed.

File: packages/nextly/src/domains/document-lock/contract.ts [hunk 6]
```diff
@@ -0,0 +1,28 @@
...
+ * drifts from the first the moment one is tuned.
+ *
+ * @module domains/document-lock/contract
+ */
+
+export {
+  DOCUMENT_LOCK_HEARTBEAT_INTERVAL_MS,
+  DOCUMENT_LOCK_LOSS_AFTER_MS,
+  DOCUMENT_LOCK_TTL_SECONDS,
+} from "./timings";
```
[the remark this answers is not in the document]
Comment (mobeenabdullah) [reply packages/nextly/src/domains/document-lock/contract.ts:21]: Confirmed by tracing the built graph rather than the declarations, and you were right: `dist/domains/document-lock/contract.mjs` reached `drizzle-orm` through `timings.ts` -> `database/lease-clock.ts`, whose `import { sql, type SQL }` is a value import. The `.d.ts` surface was clean and that is exactly why this was missed; it says nothing about what loads.

`deriveLeaseTimings` and `LeaseTimings` now live in `database/lease-timings.ts`, which imports nothing, and `lease-clock` keeps only the three functions that build SQL. The rebuilt entry resolves no external package at all.

Added `pnpm check:client-entry-graph`, wired into CI: for every `"use client"` module in the admin, the published subpath it imports must reach no database package. It walks the SOURCE graph, not `dist`, because a bundler can split a leak into a conditionally-loaded chunk and a `dist`-reading check passes whenever `dist` is stale. It reports an unresolved import rather than treating it as reaching nothing. 425 client modules, 15 nextly imports, and it fails when the old import is restored.

File: packages/admin/src/hooks/queries/useDocumentLock.ts (hunk 5 above)
[the remark this answers is not in the document]
Comment (mobeenabdullah) [reply packages/admin/src/hooks/queries/useDocumentLock.ts]: Fixed. The effect sets `acquiring` before asking, so a document changing under a mounted editor cannot leave the previous one's answer on screen. Added `acquiring` to the state union alongside `unavailable` and `lost`. Test re-renders with a new `entryId` against a pending POST and asserts the state is `acquiring`, not the old `held-by-other`.

File: packages/admin/src/hooks/queries/useDocumentLock.ts [hunk 7]
```diff
@@ -0,0 +1,163 @@
...
+        .patch<MutationEnvelope<RenewDocumentLockOutcome>>("/document-lock", {
+          ...ref,
+          claimToken: claimToken.current,
+        })
+        .then(({ item }) => {
+          if (run !== generation.current || item.status === "renewed") return;
+          // Taken over. Stop renewing and say who has it: asking again would
+          // either fail forever or re-take a claim just reported as lost.
+          claimToken.current = null;
+          setState({ status: "taken-over", holder: item.holder });
```
[the remark this answers is not in the document]
Comment (mobeenabdullah) [reply packages/admin/src/hooks/queries/useDocumentLock.ts]: Fixed, and this was the sharpest of the set. A takeover replaces the token WITHIN one effect run, so a generation check cannot separate the two overlapping heartbeats and the displaced token's `lost` answer cleared the claim the takeover had just won. Each PATCH now captures the token it sent and drops its reply unless that token is still the current one. Test: acquire, take over to t1, start a renewal, take over again to t2, then settle the first renewal as `lost` - the state stays `held-by-me`. Mutating the fence back to a run check fails it.

File: packages/admin/src/hooks/queries/useDocumentLock.ts [hunk 8]
```diff
@@ -0,0 +1,163 @@
...
+      setState({ status: "idle" });
+      return;
+    }
+
+    generation.current += 1;
+    const run = generation.current;
+    void claim(false, run);
+
+    const heartbeat = setInterval(() => {
+      if (!claimToken.current || run !== generation.current) return;
```
[the remark this answers is not in the document]
Comment (mobeenabdullah) [reply packages/admin/src/hooks/queries/useDocumentLock.ts]: Fixed by the same path. When there is no token the beat re-acquires rather than returning, so a refused claim is re-asked and the editor learns the holder has gone. A plain acquire never steals a live claim - `isTakeable` requires lapsed, own, or explicit takeover - so polling this way is safe. The holder reading is compared before it is written to state, so the poll does not re-render the editor every 15s for no news; there is a test for that too.

**Comment by mobeenabdullah** [comment]
@codex review

File: scripts/check-client-entry-graph.mjs [hunk 9]
```diff
@@ -0,0 +1,272 @@
...
+/** Where a browser bundle imports something that can reach a database. */
+export function findViolations(files, unsafeEntries) {
+  const violations = [];
+  let importSites = 0;
+  for (const [file, source] of files) {
+    for (const specifier of parseValueImports(source)) {
+      if (specifier !== "nextly" && !specifier.startsWith("nextly/")) continue;
+      importSites += 1;
+      const reach = unsafeEntries.get(specifier);
+      if (reach) violations.push({ file, specifier, reach });
```
[the remark this answers is not in the document]
Comment (mobeenabdullah) [reply scripts/check-client-entry-graph.mjs]: Confirmed and fixed, and you were right about the scale of it. Measured on this tree: the `"use client"` closure is 1038 modules against the 425 files carrying the directive, so the check was blind to the majority of the bundle it claimed to cover. I verified the specific case by making a non-client module import `nextly/runtime` - the old version passed it.

The walk is now the closure, starting from every client entry and crossing into workspace packages by resolving their export maps back to source. It reports the import chain that reached the driver, grouped by package with the shortest path to each, because one leaked entry reaches hundreds of modules across a handful of drivers and printing every path buries the import that needs changing.

Two resolution gaps surfaced while doing it, both of which had been silently stopping the walk: NodeNext `.js` specifiers (`./banner.js` beside `banner.ts`) and the nested export-condition shape `{ import: { types, default } }` that `@nextlyhq/ui` uses where `nextly` uses the flat one. Both are handled, and an import that still cannot be resolved fails the run rather than counting as reaching nothing.

On the resolver-vs-scan point specifically: I also replaced the regex with the TypeScript AST, which found a defect the scan already had. `components/features/entries/APIPlayground/generate-code.ts` prints `import { getNextly } from "nextly"` as example code for the reader, and the pattern read that string as a real import - a false positive against this very rule. The parse also erases `import type` while keeping `import { a, type B }`, sees dynamic `import()` at all, and settles the directive without a comment-stripping pass of its own.

I considered dependency-cruiser rather than extending this. It expresses reachability rules well, but its `from` selector is path-based and the entry set here is defined by a source directive, so it would need the same parse in front of it; and the repository's convention is small tested check scripts. 43 tests, and nine mutations of the rule all fail it.

File: packages/admin/src/hooks/queries/useDocumentLock.ts [hunk 10]
```diff
@@ -0,0 +1,244 @@
...
+
+    const heartbeat = setInterval(() => {
+      if (cancelled) return;
+
+      if (token === null) {
+        // Nothing to confirm. Either a colleague holds it — and asking again is
+        // how this editor learns they have left, since a plain acquire never
+        // steals a live claim — or the last attempt failed and this is the
+        // retry. Once displaced or lost, the effect has stopped the beat.
+        void acquire(false);
```
[the remark this answers is not in the document]
Comment (mobeenabdullah) [reply packages/admin/src/hooks/queries/useDocumentLock.ts:291]: Fixed. Acquires serialise: a second one cannot start while one is pending. You are right about the mechanism - `isTakeable` returns true for a live claim from the same owner, so the duplicate mints a fresh token and invalidates the one in flight, and the editor can end up renewing a token the server already replaced, reported as a takeover naming themselves.

One thing I added beyond the finding: serialising alone silently drops a take-over pressed while a poll is in flight, which turns a person's decision into nothing while the poll that displaced it only asks politely. A take-over arriving during a request is queued and run when the slot frees. Both are tested, and the queue was found by a mutation surviving - the flag it cleared turned out to be redundant, and removing it exposed the dropped click.

File: packages/admin/src/hooks/queries/useDocumentLock.ts [hunk 11]
```diff
@@ -0,0 +1,244 @@
...
+          token = null;
+          clearInterval(heartbeat);
+          setState({ status: "taken-over", holder: item.holder });
+        })
+        .catch(() => {
+          if (cancelled || sent !== token) return;
+          if (Date.now() - confirmedAt < DOCUMENT_LOCK_LOSS_AFTER_MS) return;
+          token = null;
+          clearInterval(heartbeat);
+          setState({ status: "lost" });
```
[the remark this answers is not in the document]
Comment (mobeenabdullah) [reply packages/admin/src/hooks/queries/useDocumentLock.ts]: Fixed, and this is the sharper version of a deadline I had only half-implemented. The check now runs at the top of every beat, before anything is sent, so it does not depend on a request settling. A stalled connection never rejects, and the previous placement in `.catch()` meant the claim stayed `held-by-me` indefinitely while the lease expired server-side and a colleague could take the row.

Test holds a renewal in a promise that never settles and asserts `lost` is still reached at the deadline; mutating the watchdog away fails it.

File: packages/admin/src/hooks/queries/useDocumentLock.ts [hunk 12]
```diff
@@ -0,0 +1,244 @@
...
+          if (cancelled || sent !== token) return;
+          if (item.status === "renewed") {
+            confirmedAt = Date.now();
+            return;
+          }
+          // Taken over. Stop asking: a further acquire would re-take a claim the
+          // editor was just told they had lost.
+          token = null;
+          clearInterval(heartbeat);
+          setState({ status: "taken-over", holder: item.holder });
```
[the remark this answers is not in the document]
Comment (mobeenabdullah) [reply packages/admin/src/hooks/queries/useDocumentLock.ts]: Fixed. The interval is now cleared only by cleanup. Losing the claim sets a flag the beat reads and returns on, and a successful acquire clears that flag, so taking the document back resumes renewal on the interval that was still running.

You are right that the previous shape was worse than not locking: the editor is told they hold it while nothing renews, and it expires one lease later without warning. Test takes the document back after a takeover and asserts the next beat renews with the new token; mutating `clearInterval` back into the loss path fails it.

**Comment by mobeenabdullah** [comment]
@codex review

File: packages/admin/src/hooks/queries/useDocumentLock.ts [hunk 13]
```diff
@@ -0,0 +1,303 @@
...
+    setState({ status: "acquiring" });
+
+    // Best effort. The claim lapses on its own, so a release that never arrives
+    // costs the next editor a wait rather than access.
+    const release = (claimToken: string) =>
+      void protectedApi
+        .delete("/document-lock", { ...ref, claimToken })
+        .catch(() => undefined);
+
+    const acquire = async (takeover: boolean) => {
```
[the remark this answers is not in the document]
Comment (mobeenabdullah) [reply packages/admin/src/hooks/queries/useDocumentLock.ts:203]: Fixed, and this was the finding that actually failed the run rather than warned. `acquire` is split into the two ways it can end (`installAcquired` and `installHeld`), and the walk in the new boundary test is split so deciding what one specifier means is its own step.

Verified with the tool rather than by estimating: `fallow audit --gate new-only --changed-since origin/main` now reports `verdict: pass` with `complexity_introduced: 0`, where it reported two introduced findings before.

File: packages/admin/src/hooks/queries/useDocumentLock.ts [hunk 14]
```diff
@@ -0,0 +1,303 @@
...
+        // every later beat would exit at its token guard, so an editor would
+        // spend the whole session silently unprotected. The beat retries.
+        acquiring = false;
+        if (!cancelled) setState({ status: "unavailable" });
+        drainQueuedTakeover();
+        return;
+      }
+      acquiring = false;
+
+      drainQueuedTakeover();
```
[the remark this answers is not in the document]
Comment (mobeenabdullah) [reply packages/admin/src/hooks/queries/useDocumentLock.ts]: Fixed, and the sequence you describe is exactly right: the holder leaves while the polite poll is open, the poll returns `acquired`, and draining first fires a second acquire against a claim this editor has just won. If that one rejects, the interface reports `unavailable` over a token that is held and still renewing, and no later renewal puts it back.

The outcome is installed first, and the queue is drained only when the answer was a refusal. Winning also spends the queued request, since a take-over that already got what it asked for must not be carried into a later claim and displace a colleague nobody asked to displace - that second part came out of a mutation surviving, which showed the line clearing the flag was untested.

`acquire` is split into the two ways it can end, which also takes it back under the complexity gate that was failing this PR.

File: packages/admin/src/hooks/queries/useDocumentLock.ts [hunk 15]
```diff
@@ -0,0 +1,303 @@
...
+    const drainQueuedTakeover = () => {
+      if (!queuedTakeover || cancelled) return;
+      queuedTakeover = false;
+      void acquire(true);
+    };
+
+    /** Give the claim up locally, without asking the server for anything. */
+    const surrender = (next: DocumentLockState) => {
+      token = null;
+      surrendered = true;
```
[the remark this answers is not in the document]
Comment (mobeenabdullah) [reply packages/admin/src/hooks/queries/useDocumentLock.ts]: Fixed. `surrender` releases before forgetting. You are right about the mechanism: the renewal whose reply never arrived may still have reached the server and extended the lease by most of a TTL, so colleagues went on seeing this editor as the holder long after its own interface said otherwise.

The delete is token-scoped (`WHERE id = ... AND claim_token = ...`), so it removes nothing when the claim really has moved on, which is why doing it for a server-reported takeover as well as a local deadline is safe rather than a second decision. Test holds a renewal in a promise that never settles, advances past the deadline and asserts the DELETE.

File: scripts/check-client-entry-graph.mjs [hunk 16]
```diff
@@ -0,0 +1,373 @@
...
+        ts.isStringLiteral(node.moduleSpecifier)
+      ) {
+        specifiers.push(node.moduleSpecifier.text);
+      }
+    } else if (
+      ts.isCallExpression(node) &&
+      node.expression.kind === ts.SyntaxKind.ImportKeyword &&
+      node.arguments.length > 0 &&
+      ts.isStringLiteral(node.arguments[0])
+    ) {
```
[the remark this answers is not in the document]
Comment (mobeenabdullah) [reply scripts/check-client-entry-graph.mjs]: You are right, and the pointer to the repository's own reader was the more useful half of the finding. `require("...")` and `import x = require("...")` produced no specifier at all, so the gate could report a clean closure over a bundled driver.

The hand-written visitor is gone. It now uses `@nextlyhq/module-specifiers`, which already covers those two forms plus `typeof import`, JSDoc `@import`, triple-slash references and a dynamic target it cannot read (reported as `UNRESOLVABLE_SPECIFIER`, which this treats as a violation rather than as nothing). Duplicating that reader was the exact defect `.claude/rules/derived-checks.md` forbids, and the reader exists because three packages had already grown their own copy.

One thing that reader could not yet answer: it collects type-only references deliberately, and its docblock says a caller wanting runtime reachability "filters afterwards" - which a list of strings does not permit. Measured, that matters here: unfiltered, the same walk reports 2072 modules and ten reachable database packages, every one behind an `import type`, and an erased reference cannot put code in a bundle. So `moduleSpecifierRefs` now reports a `typeOnly` flag per reference and `importedSpecifiers` is derived from it, one walk so the two cannot drift. A mixed clause `import { a, type B }` is deliberately NOT type-only.

The check also moved: it is a test in the package that owns the boundary, beside `packages/ui/src/layering.test.ts`, which does the same kind of graph check for the same reason. That is what lets it import the workspace reader at all, and `pnpm test` already runs it, so the separate CI step and the root script are both removed. Synthetic cases feed the same functions trees that MUST fail - a driver seven hops past the directive, a `require`, an `import =`, a dynamic import, and an unreadable target - because an empty result is not evidence on its own.

**Comment by mobeenabdullah** [comment]
@codex review

File: packages/admin/src/hooks/queries/useDocumentLock.ts [hunk 17]
```diff
@@ -0,0 +1,323 @@
...
+      const sent = token;
+      void protectedApi
+        .patch<MutationResponse<RenewDocumentLockOutcome>>("/document-lock", {
+          ...ref,
+          claimToken: sent,
+        })
+        .then(({ item }) => {
+          if (cancelled || sent !== token) return;
+          if (item.status === "renewed") {
+            confirmedAt = Date.now();
```
[the remark this answers is not in the document]
Comment (mobeenabdullah) [reply packages/admin/src/hooks/queries/useDocumentLock.ts:304]: Fixed in #1590. Both paths now install the dispatch timestamp: the lease starts when the server processes the claim, so timing it from the reply credited this editor with however long that reply spent in transit, and a reply delayed past the margin left the hook reporting a claim that had already expired server-side.

Two tests, one per path, each built so the two readings disagree: the request is sent on one beat and answered much later, and the deadline that follows is reached only if the send time is what counted. Mutating either back to `Date.now()` at receipt fails them.

File: packages/module-specifiers/src/index.ts [hunk 18]
```diff
@@ -104,41 +144,55 @@ export function importedSpecifiers(text: string, fileName: string): string[] {
...
+        specifier: ts.isStringLiteralLike(target)
+          ? target.text
+          : UNRESOLVABLE_SPECIFIER,
+        typeOnly: false,
+      });
     } else if (ts.isCallExpression(node)) {
       const callee = node.expression;
       const resolvesAModule =
         callee.kind === ts.SyntaxKind.ImportKeyword ||
         (ts.isIdentifier(callee) && callee.text === "require");
```
[the remark this answers is not in the document]
Comment (mobeenabdullah) [reply packages/module-specifiers/src/index.ts:186]: Fixed in #1590, and the pointer to this repository's own corpus was the useful part: `packages/ui/src/layering.test.ts` already treats `module.require` as a runtime resolver, so the two readers disagreed and the newer one was the weaker.

`@nextlyhq/module-specifiers` now recognises it, both spellings, since `a.b` and `a["b"]` are the same read and a rule for one is a rule the other walks around. Deliberately narrow: the receiver has to be `module` itself, because treating every `.require` as a resolve reports ordinary code as a dependency nobody has - there is a negative control for `loader.require` and `holder["require"]`. The boundary test's CommonJS case covers both new spellings too.

File: packages/admin/src/hooks/queries/useDocumentLock.ts [hunk 19]
```diff
@@ -0,0 +1,323 @@
...
+        holder = current;
+        setState({ status: "held-by-other", holder: current });
+      }
+    };
+
+    const acquire = async (takeover: boolean) => {
+      if (acquiring) {
+        if (takeover) queuedTakeover = true;
+        return;
+      }
```
[the remark this answers is not in the document]
Comment (mobeenabdullah) [reply packages/admin/src/hooks/queries/useDocumentLock.ts:207]: Fixed in #1590. You are right that this is not the rejection path: nothing cleared the one-at-a-time slot, so every later beat returned at the serialisation guard and the editor sat in `acquiring` for the whole session with a take-over queued behind a request that was never coming back.

`fetcher` spreads `RequestInit`, so claims now carry an `AbortSignal` bounded by one heartbeat - a claim that has not answered within a beat cannot help that beat - and the next beat retries. The test mocks the signal the way `fetch` behaves, since that abort is the mechanism under test.

On the late server-side acquisition you raise: an aborted request may still have been processed, leaving a claim this editor does not know it holds. That is bounded by the lease and reclaimed by the next acquire, since the server treats the same owner as takeable. It is stated in the code rather than left to be discovered.

File: packages/admin/src/hooks/queries/useDocumentLock.ts [hunk 20]
```diff
@@ -0,0 +1,323 @@
...
+
+    /** Store a refusal, quietly when the holder has not changed. */
+    const installHeld = (current: DocumentLockHolder) => {
+      token = null;
+      // This runs every beat while a colleague holds the document, and a fresh
+      // object each time re-renders the editor for no news.
+      if (holder === null || !sameHolder(holder, current)) {
+        holder = current;
+        setState({ status: "held-by-other", holder: current });
+      }
```
[the remark this answers is not in the document]
Comment (mobeenabdullah) [reply packages/admin/src/hooks/queries/useDocumentLock.ts:200]: Fixed in #1590. The cached holder is cleared on failure. Exactly as you describe, an identical reading on the next successful poll was suppressed as "nothing changed" and the editor stayed on `unavailable` while the server answered perfectly well - and identical is the likely case, since the holder renews on this same cadence.

Test fails the poll, then returns the same holder with the same `expiresInSeconds`, and asserts the editor goes back to `held-by-other`.

File: packages/admin/src/hooks/queries/useDocumentLock.ts [hunk 21]
```diff
@@ -0,0 +1,323 @@
...
+        return;
+      }
+      acquiring = false;
+
+      if (cancelled) {
+        // 🔴 The claim outlived the editor that asked for it: this reply landed
+        // after cleanup, which found no token to release. Releasing it here is
+        // the difference between a colleague waiting one request and waiting a
+        // whole lease.
+        if (item.status === "acquired") release(item.claimToken);
```
[the remark this answers is not in the document]
Comment (mobeenabdullah) [reply packages/admin/src/hooks/queries/useDocumentLock.ts:230]: Fixed in #1590, and this one taught me something about the two earlier fences. A late reply from a cleaned-up run displaces the run that replaced it, because the server treats a claim from the same owner as takeable - so releasing it left the live run holding a token the server had already forgotten.

The displaced run is now told to claim again, through a signal the live run publishes. What it did NOT need is a second generation counter beside the token fence: React runs a cleanup before the effect that replaces it, so every superseded run is already cancelled, and the signal is refused by a run that has itself been cleaned up. I wrote that extra check first and removed it when a mutation showed nothing could fail on it.
